# Chapter 1

In [ ]:
import nltk
print(nltk.__version__) 

In [ ]:
#download - nltk download
nltk.download()

In [ ]:
#sent_tokenizer

#کلمات مخفف- هم مکان ها و کلماتی که در شروع جمله ها میایند یک مدل میسازه. unsupervised - روی متن بزرگ
# it's instance of "PunktSentenceTokenizer" . ""everytime train model""

# A1
from nltk.tokenize import sent_tokenize
a=sent_tokenize('Hi students! Welcome to our course.')

#A2
import nltk.data
tokenizer = nltk.data.load('tokenizers/punkt/PY3/english.pickle')
tokenizer.tokenize('Hi students! Welcome to our course.')

# #hazm #sent_tokenize
from hazm import sent_tokenize
sent_tokenize('سلام چطوری؟ اسم تو بگو توپولی')

In [ ]:
# #word_tokenizer #word_tokenize


para = "he can't do it!"

# A1 - TreebankWordTokenizer
from nltk.tokenize import word_tokenize

treebank = word_tokenize(para)

# A2 - TreebankWordTokenizer
from nltk.tokenize import TreebankWordTokenizer

tokenizer = TreebankWordTokenizer()
treebank=tokenizer.tokenize(para)

#### تری بنک یک مدل از پیش آماده نیست . و بر اساس مدل استخراج شده از یک کورپس به همین اسم و با استفاده از
#### رجکس ساخته شده




#### #WordPunktTokenizer

from nltk.tokenize import WordPunctTokenizer

tokenizer = WordPunctTokenizer()
wordPunkt=tokenizer.tokenize("can't is a contraction.")



#### #hazm #word_tokenize
from hazm import word_tokenize

hazmWordTokenize = word_tokenize('سلام سلام چطوری اسم تو بگو توپولی')

treebank,wordPunkt,hazmWordTokenize

In [1]:
# #sent_tokenize #Tokenizing #sentences #regex #regular #word_tokenize 

# برای زمانی که بخواهیم کنترل کامل روی توکن کردن متن داشته باشیم.

# RegexTokenizer => TokenizerI  --> why use "RegexTokenizer" instead of "re.findall()" ?
# use another pars of nltk package - for exp : corpus reader oprtional tokenizer 
# must be an instance of TokenizerI


# A1
from nltk.tokenize import RegexpTokenizer

t = RegexpTokenizer(r"[\w']+")
t.tokenize("can't is a contraction.")

# A2
from nltk.tokenize import regexp_tokenize

regexp_tokenize("can't is a contraction.", "[\w']+")



["can't", 'is', 'a', 'contraction']

In [ ]:
# #sent_tokenize #sentence #tokenize #train

# PunktSentenceTokenizer از طریق یادگیری بدون نظارت یادمیگیره چجوری جملات رو بشنکنه
# مثلا اگر یک توکن معمولا با نقطه بیاید احتمالا خلاصه است.
# mr.amini


from nltk.tokenize import PunktSentenceTokenizer
from nltk.tokenize import sent_tokenize
from nltk.corpus import webtext


text = webtext.raw('overheard.txt')
sent_tokenizer = PunktSentenceTokenizer(text)

pre_trained = sent_tokenize(text)
our_trained = sent_tokenizer.tokenize(text)


# A2

with open('~/nltk_data/corpora/webtext/overheard.txt',encoding='ISO-8859-2') as f:
    text = f.read()
    
sent_tokenizer = PunktSentenceTokenizer(text)


In [ ]:
pre_trained[678],our_trained[678]

In [ ]:
#stopword
from nltk.corpus import stopwords
stopwords.words('english')
stopwords.fileids() # for #corpus

In [ ]:
#wordnet #synsets

from nltk.corpus import wordnet as wn

syns = wn.synsets('book') #list of synset

wn.synset('book.n.01') == syns[0]

syns[0].name(),syns[0].definition(),syns[0].examples()

In [ ]:
#wordnet #hypernyms = فراشمول یا انتزاعی #hyponyms = زیرشمول

from nltk.corpus import wordnet as wn

syn = wn.synset('book.n.01')

syn.hypernyms()[0].hyponyms()

syn.root_hypernyms() # ریشه درخت - انتزاعی ترین حالت 

syn.hypernym_paths() # لیستی از لیست ها - مسیر تا ریشه که ممکن از چند مسیر باشد

# ----------------------------------------------------

#similarity #shortest_path #common

s1 = wn.synsets('car')[0]
s2 = wn.synsets('computer')[0]

##########  شباهت یابی - راه 1

#اگر هایپرنیم های مشترک نداشته باشند نمیتوان با استفاده ازین محاسبه کرد  و  نان برمیگرذاند

s1.wup_similarity(s2) # یک عدد بین 0 و ۱ هرچی بیشتر شبیه تر




##########راه 2 - هر چی فاصله بیشتر کم تر شبیه اند

#هایپرنیم های دو سینست را که یکسان اند تا ریشه بدست می آورد
#s1.hypernym_paths() and s2....

s1.common_hypernyms(s2)



########## راه 3

s1.path_similarity(s2)


########## راه 4

#leacock chordorow

s1.lch_similarity(s2)  ## هر چی بیشتر شبیه تر -- محدوده ای براشون پیدا نکردم.

In [ ]:
# wordnet #part of speech #pos 

from nltk.corpus import wordnet as wn

syn = wn.synset('book.n.01')

syn.pos() #notice - pos just for "synset" not synsets

# -----------------------------------

syns = wn.synsets('book',pos='n') # just nouns

In [ ]:
#wordnet #lemma #antonyms

from nltk.corpus import wordnet as wn

s = wn.synset('cookbook.n.01')

s.lemmas() #all lemma

s.lemmas()[0].name(),s.lemmas()[0].synset()

# ------------------------------------------


s = wn.synset('good.a.01')
s.lemmas()[0].antonyms()[0]  # "some" lemma has antonyms : return lemma

In [ ]:
#collocations #bigram #bigramcollocationsfinder

from nltk.corpus import webtext
from nltk.collocations import BigramCollocationFinder
from nltk.metrics import BigramAssocMeasures
from nltk.corpus import stopwords

words = [word.lower() for word in webtext.words('grail.txt')]

bcf = BigramCollocationFinder.from_words(words)

bcf.apply_word_filter(lambda w:len(w)<3 or w in stopwords.words('english'))

bcf.nbest(BigramAssocMeasures.likelihood_ratio,4)


#trigram #trigramcollocations

from nltk.collocations import TrigramCollocationFinder
from nltk.metrics import TrigramAssocMeasures

words = [word.lower() for word in webtext.words('singles.txt')]
tcf = TrigramCollocationFinder.from_words(words)

tcf.apply_freq_filter(3) #باید سه بار یا بیشتر تکرار شده باشند تا کولوکیشن حساب شوند 
tcf.apply_word_filter(lambda x:len(x)<3 or x in stopwords.words('english'))




#-------------------------------------------------------------

tcf.nbest(TrigramAssocMeasures.likelihood_ratio,4) #score_fn,n    n top scores

tcf.score_ngrams(TrigramAssocMeasures.likelihood_ratio)  #score all ngrams

tcf.above_score(TrigramAssocMeasures.likelihood_ratio,5) #score_fn,min_score   all ngrams with greater min_score

# Chapter 2

In [18]:
#stemming #stemmer #lancaster #regexp #snowball

from nltk.stem import PorterStemmer,LancasterStemmer,RegexpStemmer,SnowballStemmer

#porter
stemmer = PorterStemmer()
stemmer.stem('cookery')

#lancaster 
stemmer = LancasterStemmer()
stemmer.stem('cookery')

#regexp
stemmer = RegexpStemmer('ing')
stemmer.stem('booking')

#snowball
SnowballStemmer.languages
stemmer = SnowballStemmer('english')
stemmer.stem('booking')

#hazm
from hazm import Stemmer
stemmer = Stemmer()
stemmer.stem('میرویم')

'میرو'

In [24]:
#lemmatize

from nltk.stem import WordNetLemmatizer

l = WordNetLemmatizer()
#default pos = noun
l.lemmatize('cooking'),l.lemmatize('cooking',pos='n'),l.lemmatize('cooking',pos='v')

('cooking', 'cooking', 'cook')

In [ ]:
#replacement pattern #contractions
import re
from nltk.tokenize import word_tokenize

rp = [
    ('(\w+)\'ll','\g<1> will'),
    ('([i,I])\'m','\g<1> am')
]

rp = [(re.compile(regex),rep) for (regex,rep) in rp]

s_in = "i'm I'll"

for (pattern,rep) in rp:
    s_in = re.sub(pattern,rep,s_in)

word_tokenize(s_in)

In [10]:
#looove to love #removing repeating character #backreference #backrefrence

import re
from nltk.corpus import wordnet

class RepeatReplacer():
    def __init__(self):
        self.pattern = re.compile(r'(\w*)(\w)\2(\w*)')
        self.rep = r'\1\2\3'
    def replace(self,word):
        if wordnet.synsets(word):
            return word
        repl_word = self.pattern.sub(self.rep,word)
        if word!=repl_word:
            return self.replace(repl_word)
        else:
            return word

a=RepeatReplacer()
a.replace('looove')

'love'

In [ ]:
#pyenchant #enchant #spell correcting #spell correction

import enchant
from nltk.metrics import edit_distance

class SpellingReplacer:
    def __init__(self,dict_name='en',max_distance=2):
        self.dict = enchant.Dict(dict_name)
        self.max_distance = max_distance
        
    def replace(self,word):
        if self.dict.check(word):
            return word
        else:
            sugs = self.dict.suggest(word)
            if sugs and edit_distance(word,sugs[0])<=self.max_distance:
                return sugs[0]
            else:
                return word

#note :

#dict = enchant.Dict(dict_name)
#dict.check(word) => True/False
#dict.suggest(word) => list of suggestion

#dict = enchant.DictWithPWL('en_US', 'my_words.txt')  ##personal word list

enchant.list_dicts(),enchant.list_languages(),enchant.dict_exists('en')

In [ ]:
#replacing synonyms #replace synonyms #yaml #csv

import csv
import yaml

class WordReplacer(object):
    def __init__(self, word_map):
        self.word_map = word_map
    def replace(self, word):
        return self.word_map.get(word, word)


class CsvWordReplacer(WordReplacer):
    def __init__(self, fname):
        word_map = {}
        for line in csv.reader(open(fname)):
            word, syn = line
            word_map[word] = syn
        super().__init__(word_map)
        

class YamlWordReplacer(WordReplacer):
    def __init__(self, fname):
        word_map = yaml.load(open(fname))
        super().__init__(word_map)
        
## yaml file example:=>  
## name : xxx

In [ ]:
#negations #Replacing negations with antonyms

from nltk.corpus import wordnet
from nltk.tokenize import word_tokenize

class AntonymReplacer(object):
    def replace(self, word, pos=None):
        antonyms = set()
        for syn in wordnet.synsets(word, pos=pos):
            for lemma in syn.lemmas():
                for antonym in lemma.antonyms():
                    antonyms.add(antonym.name())
        if len(antonyms) == 1: # unambiguous replacement
            return antonyms.pop()
        else:
            return None
        
    def replace_negations(self, orig_sent):
        sent = word_tokenize(orig_sent)
        i, l = 0, len(sent)
        words = []
        while i < l:
            word = sent[i]
            if word == 'not' and i+1 < l:
                ant = self.replace(sent[i+1])
                if ant:
                    words.append(ant)
                    i += 2
                    continue
            words.append(word)
            i += 1
        return words

class AntonymWordReplacer(WordReplacer, AntonymReplacer):
    pass

# Chapter 4

In [ ]:
#POS tagging #tag #tagger #untagging #unigram #bigram #trigram #quadgram

from nltk.corpus import treebank

sents = treebank.sents() 

# all tags : http://www.ling.upenn.edu/courses/Fall_2003/ling001/penn_treebank_pos.html

tagged_sents = treebank.tagged_sents() # [ [('Pierre', 'NNP'),('Vinken', 'NNP'),...],... ]


#--------------------------------------------------

#default tagger

# چون تمامی این تگر ها از کلاس پدر یکسانی ارث بری میکنند توابع زیر بینشان یکسان است

from nltk.tag import DefaultTagger

tagger = DefaultTagger('NN')

tagger.tag(sents[0]) #one sentence
tagger.tag_sents(sents) #list of sentences
tagger.tag_sents([['you','are','beautiful'],['shame','on','you']])

test_sents = tagged_sents[3000:]
tagger.evaluate(test_sents)  ## تگر آموزش دیده را بهش دیتا لیبل دار دادیم تا تست کنه خودشو

#---------------------------------------------------

#UnigramTagger(train=None, model=None,backoff=None, cutoff=0, verbose=False)

#cutoff : موارد زیر این تعداد تکرار را در نظر نمیگیرد و با آن ها مدل نمیسازد
#model : prepared model -> model={'Pierre': 'NN'}
#backoff : اگر تگر نتونست تگ بزنه میره سراغ این یکی

from nltk.tag import UnigramTagger

tagger = UnigramTagger(train = tagged_sents[:3000])

#tagger.tag(sents[0])
#tagger.tag(sents)

tagger.evaluate(tagged_sents[3000:])

#---------------------------------------------------

#NgramTagger(n, train=None, model=None,backoff=None, cutoff=0, verbose=False)

from nltk.tag import NgramTagger

class QuadgramTagger(NgramTagger):
    def __init__(self,train=None, model=None,backoff=None, cutoff=0, verbose=False):
        NgramTagger.__init__(self,4,train,model,backoff,cutoff,verbose)


#---------------------------------------------------

#backoff tagging

from nltk.tag import UnigramTagger,BigramTagger,TrigramTagger

# نکته : اول تریگرام هارو میزنه بعد بایگرام ها بعد ...
def backoff_tagger(train_sents,tagger_classes,backoff=None):
    for tagger_class in tagger_classes:
        backoff = tagger_class(train=train_sents,backoff = backoff)
    return backoff


tagger=backoff_tagger(tagged_sents[:3000],[UnigramTagger,BigramTagger,TrigramTagger,QuadgramTagger],DefaultTagger('NN'))

tagger._taggers #[<TrigramTagger: size=845>,<BigramTagger: size=1859>,<UnigramTagger: size=8818>,<DefaultTagger: tag=NN>]

tagger.evaluate(tagged_sents[3000:])

tagger._context_to_tag 


#---------------------------------------------------
#save tags #store tags #save model

#import pickle
# f = open('tagger.pickle', 'wb')
# pickle.dump(tagger, f)
# f.close()
# f = open('tagger.pickle', 'rb')
# tagger = pickle.load(f)

#---------------------------------------------------

#Untagging a tagged sentence

from nltk.tag import untag

untag([('Hello', 'NN'), ('students', 'NN')])

In [42]:
#FreqDist #ConditionalFreqDist #?



In [56]:
#Create custome model with FreqDist and Conditional FreqDist

#ساخت یک مدل از ۱۰۰۰ کلمه ای که تقریبا همیشه یک تگ یکسان دارند.

from nltk.corpus import treebank
from nltk.probability import FreqDist,ConditionalFreqDist


def word_tag_model(words,tagged_words,limit=1000):
    fd = FreqDist(words)
    cfd = ConditionalFreqDist(tagged_words)
    
    # cfd['book'] ==> FreqDist({'NN': 7, 'VB': 1})
    # fd.most_common(2) ==> [(',', 4885), ('the', 4045)]
    
    most_common = [word for word,count in fd.most_common()]
    return {word:cfd[word].max() for word in most_common}

model=word_tag_model(treebank.words(),treebank.tagged_words())

tagger=backoff_tagger(tagged_sents[:3000],[UnigramTagger,BigramTagger,TrigramTagger,QuadgramTagger],DefaultTagger('NN'))

custom_tagger = UnigramTagger(model=model,backoff=tagger)

custom_tagger.evaluate(tagged_sents[3000:])

0.9632203755665876

In [62]:
#tagging with regex #regex

from nltk.tag import RegexpTagger

#این ها به ترتیب از بالا به پایین چک می شوند. الگوری اخری برای زمانی است که هیچ کدام از 
# بالایی ها چک نشده باشند.

patterns = [
    (r'^\d+$','CD'),
    (r'.*ing$','VBG'),
    (r'.*ful$', 'JJ'),
    (r'.*','NN')
]

tagger = RegexpTagger(patterns)
tagger.evaluate(tagged_sents[3000:])

0.16861644722641916

In [ ]:
#affix tagger #prefix #suffix

from nltk.tag import AffixTagger

#AffixTagger(train=None, model=None, affix_length=-3,min_stem_length=2, backoff=None, cutoff=0, verbose=False)

#min_stem_length =>>>  |affix_length| + min_stem_length <= len(word)  else==> return None
#affix_length =>>> if pos-> word[:affix_length], if neg ->  word[affix_length:]

# اگر -2 بود دو کاراکتر اخر و اگر +2 بود دو کاراکتر اول را یاد میگیرد.


# order for backoff => first check : -3 -> -2 -> 2 -> 3


#--------------------------------------------------------

tagger = AffixTagger(train=tagged_sents[:3000],min_stem_length=0,affix_length=0)  # equal to UnigramTagger

tagger.evaluate(tagged_sents[3000:])

#tagger._context_to_tag

In [1]:
#Brill Tagger #brill trainer  #?


from nltk.tag import brill,brill_trainer
from nltk.tag import DefaultTagger
from nltk.corpus import treebank

tagged_sents = treebank.tagged_sents()

templates = [
    brill.Template(brill.Word([1,2]),brill.Word([-1])), # words
    brill.Template(brill.Pos([1,2]),brill.Pos([-1]))
]

trainer = brill_trainer.BrillTaggerTrainer(DefaultTagger('NN'),templates,deterministic=True)
tagger = trainer.train(tagged_sents[:3000])

tagger.evaluate(tagged_sents[3000:])

In [7]:
#tnt #tnt tagger #second order Markov models

from nltk.tag import tnt
from nltk.tag import DefaultTagger

# tnt.TnT(unk=None, Trained=False, N=1000, C=False) 

#unk -> tagger for unkown words
#if unk is pretrained Trained=True else => False  ---> notice : for all sequentialBackoffTagger set True.
#C : True if you want capitalization of words to be significant

tagger=tnt.TnT(DefaultTagger('NN'),Trained=True)
tagger.train(tagged_sents[:3000])
tagger.evaluate(tagged_sents[3000:])

0.892467083962875

In [ ]:
#wordnet #tagging with wordnet #wordnet tagging

#v -> VB   verb
# n -> NN  noun
# a -> JJ  adjective
# s -> JJ  adjective
# r -> RB  adverb



#دقت کن . فقط دو تااز توابع رو ارث بری کرده و برای کار کافی است.
# به تنهایی نتیجه ی خوبی ندارد ولی به صورت بک آف برای یونیگرام و بایگرام و... خیلی مناسب است.

from nltk.corpus import wordnet
from nltk.tag import SequentialBackoffTagger
from nltk.probability import FreqDist

class WordNetTagger(SequentialBackoffTagger):
    '''
     >>> wt = WordNetTagger()
     >>> wt.tag(['food', 'is', 'great'])
     [('food', 'NN'), ('is', 'VB'), ('great', 'JJ')]
     '''
    def __init__(self, *args, **kwargs):
        SequentialBackoffTagger.__init__(self, *args, **kwargs)
        self.wordnet_tag_map = {
            'n': 'NN', 
            's': 'JJ', 
            'a': 'JJ', 
            'r': 'RB', 
            'v': 'VB' 
        }
    def choose_tag(self, tokens, index, history):
        word = tokens[index]
        fd = FreqDist()
        for synset in wordnet.synsets(word):
            fd[synset.pos()] += 1
        if len(fd)==0:
            return 'NN'
        return self.wordnet_tag_map.get(fd.max())

In [49]:
#tagging proper names #SequentialBackoffTagger

from nltk.corpus import names
from nltk.tag import SequentialBackoffTagger

class NamesTagger(SequentialBackoffTagger):
    def __init__(self,*args,**kwargs):
        SequentialBackoffTagger.__init__(self,*args,**kwargs)
        self._name_set = set([w.lower() for w in names.words()])
    def choose_tag(self, tokens, index, history):
        if tokens[index].lower() in self._name_set:
            return 'NNP'
        return None
    
nt = NamesTagger()
nt.tag(['James', 'book'])

[('James', 'NNP'), ('book', None)]

In [50]:
#ClassifierBasedTagger #most_accurate_tagger #most accurate # معمولا بهترین نتیجه را دارد

<br><br>
<font size="8" color="red">معمولا بهترین تگر هست</font>
<br><br><br>
<font size="4">The ClassifierBasedPOSTagger class is a subclass of ClassifierBasedTagger that implements a feature detector. The feature detector finds multiple length suffixes, does some regular expression matching, and looks at the unigram, bigram, and trigram history to produce a fairly complete set of features for each word.
<br>
It only implements a feature_detector() method. All the training and tagging is done in ClassifierBasedTagger. It defaults to training a NaiveBayesClassifier class. To use other classifiers, pass in your own classifier_builder function:</font>

In [ ]:
#ClassifierBasedTagger #most_accurate_tagger #most accurate # معمولا بهترین نتیجه را دارد

from nltk.classify import MaxentClassifier
from nltk.tag import ClassifierBasedPOSTagger
from nltk.corpus import treebank

train_sents = treebank.tagged_sents()

me_tagger = ClassifierBasedPOSTagger(train=train_sents[:3000],classifier_builder=MaxentClassifier.train)
me_tagger.evaluate(train_sents)

# custome feature detector 

from nltk.tag.sequential import ClassifierBasedTagger

def unigram_feature_detector(tokens, index, history):
    return {'word':tokens[index]} # shuld return a dictionary of feature-name:feature-value

tagger = ClassifierBasedTagger(train=train_sents, feature_detector=unigram_feature_detector)
tagger.evaluate(test_sents)



# with backoff  :  cutoff_prob اگر این رو نزاریم
# استفاده از بک آف فایده ای نداره چون این همیشه بهترین رو برمیگردونه
default = DefaultTagger('NN')
tagger = ClassifierBasedPOSTagger(train=train_sents, cutoff_prob=0.3, backoff=default)
tagger.evaluate(test_sents)

# Training a tagger with NLTK-Trainer
# انتخاب بهترین روش و پارامتر ها توسط کد نوشته شده زیر:

https://github.com/japerk/nltk-trainer

python train_tagger.py treebank --fraction 0.75 --no-pickle 

75% for train, other for test.
AffixTagger + UnigramTagger + BigramTagger + TrigramTagger.

# Chapter 7

In [5]:
#bag of words

from nltk.collocations import BigramCollocationFinder
from nltk.metrics import BigramAssocMeasures

def bag_of_words(words:list)->dict:
    return {word:True for word in words}

def bag_of_words_not_in_set(words,badwords):
    return bag_of_words(set(words)-set(badwords))

def bag_of_bigrams_words(words,score_fn=BigramAssocMeasures.chi_sq,n=200):
    bcf = BigramCollocationFinder.from_words(words)
    bigrams=bcf.nbest(score_fn,n)
    return bag_of_words(bigrams+words)


In [6]:
#naive bayes #training naive bayes classifier

from nltk.corpus import movie_reviews
from collections import defaultdict

# {'label1':[ {'hello':True,'see':True},...] , 'label2':...}
def label_feats_from_corpus(corp, feature_detector=bag_of_words):
    label_feats = defaultdict(list)
    for cat in corp.categories():
        for fileid in corp.fileids(categories = [cat]):
            featureset = feature_detector(corp.words(fileids=[fileid]))
            label_feats[cat].append(featureset)
    return label_feats


# [(feat,label),...]  ---> [({'hello':True},'label1'),...]
def split_label_feats(lfeats,split=0.75):
    test_feats = list()
    train_feats = list()
    for label,feats in lfeats.items():
        cutoff = int(len(feats)*split)
        train_feats.extend([(feat,label) for feat in feats[:cutoff]])
        test_feats.extend([(feat,label) for feat in feats[cutoff:]])
    return train_feats,test_feats

label_feats = label_feats_from_corpus(movie_reviews)

train,test = split_label_feats(label_feats)

In [3]:
#naive bayes #training naive bayes classifier

from nltk.classify import NaiveBayesClassifier

nb = NaiveBayesClassifier.train(train)



## for classification a sample :

feat = bag_of_words(['the', 'plot', 'was', 'fantastic'])

# فقط بهترین لیبل را خروجی میدهد
nb.classify(feat)

# احتمال هر لیبل را محاسبه میکند و خروجی میدهد
probs = nb.prob_classify(feat)

probs.samples() #dict_keys(['neg', 'pos'])

probs.prob('neg'),probs.prob('pos')

probs.max() # best choice for label => 'pos'

nb.most_informative_features(n=5)

nb.show_most_informative_features(n=5)

#----------------------------------------------

## accuracy(trained_classifier, test_feats)

from nltk.classify.util import accuracy

accuracy(nb,test)

Most Informative Features
             magnificent = True              pos : neg    =     15.0 : 1.0
             outstanding = True              pos : neg    =     13.6 : 1.0
               insulting = True              neg : pos    =     13.0 : 1.0
              vulnerable = True              pos : neg    =     12.3 : 1.0
               ludicrous = True              neg : pos    =     11.8 : 1.0


0.728

In [3]:
#Decision Tree #DecisionTree


<b>entropy_cutoff</b>: If the entropy of the probability distribution of label choices in the tree is greater than the entropy_cutoff value, then the tree is refined further by creating more branches. <b>Higher values of entropy_cutoff will decrease both accuracy and training time.</b>

<b>depth_cutoff</b>: The default value is 100, which means that classification may require up to 100 decisions before reaching a leaf node.

<b>support_cuttoff</b>: The minimum number of instances that are required to make a decision about a feature.


In [ ]:
from nltk.classify import DecisionTreeClassifier

dt = DecisionTreeClassifier.train(train, binary=True, 
                                  entropy_cutoff=0.8, depth_cutoff=5, support_cutoff=30)



In [6]:
#scikitlearn #scikit learn #classifier 
from nltk.classify.scikitlearn import SklearnClassifier
from sklearn.naive_bayes import MultinomialNB
from nltk.classify.util import accuracy

skc = SklearnClassifier(MultinomialNB())
skc.train(train)
accuracy(skc,test) # 0.83


### OTHER ::: دقیقا شبیه بالایی اجرا می شوند

# from sklearn.naive_bayes import BernoulliNB : 0.812
# ==>can also work with discrete values by binarizing those values, so that the final values are 1 or 0.

# from sklearn.linear_model import LogisticRegression : 0.892

# from sklearn.svm import SVC : 0.69
# ==>Support vector machine -> خود ان ال تی کی توی خودش اینو نداره

# from sklearn.svm import LinearSVC : 0.864

# from sklearn.svm import NuSVC : 0.882

0.83

In [6]:
#precision #recall

import collections
from nltk.metrics import precision,recall

def precision_recall(classifier,testfeats):
    refsets = defaultdict(set)
    testsets = defaultdict(set)
    for i,(feats,label) in enumerate(testfeats):
        refsets[label].add(i)
        observed = classifier.classify(feats)
        testsets[observed].add(i)
    precision,recall = {},{}
    for label in classifier.labels():
        precision[label] = precision(refsets[label],testsets[label])
        recall[label] = recall(refsets[label],testsets[label])
    return precision,recall

# the result for naive bayes:

# nb_precisions['pos'], nb_recalls['pos'] == (0.651595744680851, 0.98)
# nb_precisions['neg'], nb_recalls['neg'] == (0.9596774193548387, 0.476)

## The conclusion could be that there are certain common words that are biased towards the pos label, 
## but occur frequently enough in the neg feature sets to cause mis-classifications.


#-----------------------------------------------------

## USE  "from sklearn.svm import NuSVC"

# SklearnClassifier class of NuSVC weighs its features according to its own internal model. 
# Words that are more significant are those that occur primarily in a single label, and will get higher weights in the model.
# Words that are common to both labels will get lower weights, as they are less significant.

In [17]:
#high information words #show_most_informative_features() 

A <font color='green'><b>high information word</b></font> is a word that is <font color='red'><b>strongly biased towards a single classification label</b></font>.

Use show_most_informative_features() method on both the NaiveBayesClassifier class and the MaxentClassifier class. The top words are <font color='green'><b>different for both classifiers</b></font> due to how each classifier calculates the significance of each feature, and it's actually beneficial to have these different methods as <font color='green'><b>they can be combined to improve accuracy</b></font> (next recipe).

<br><br>
<font size="4">Eliminating low informative words from the training data can actually improve accuracy, precision, and recall.</font>

In [16]:
#high information words #show_most_informative_features() 

# score_fn(n_ii, (n_ix, n_xi), n_xx) : ارزش کلمه به ازای یک لیبل خاص را بررسی میکند
# n_ii : مجموع تکرار کلمه در لیبل خاص
# n_ix : مجموع تکرار کلمه در کل لیبل ها
# n_xi : مجموع کل کلمات یک لیبل
# n_xx : مجموع کل کلمات برای تمامی لیبل ها

from nltk.probability import FreqDist,ConditionalFreqDist
from nltk.metrics import BigramAssocMeasures
import collections

def high_information_words(labeled_words, score_fn=BigramAssocMeasures.chi_sq, min_score=5):
    word_fd = FreqDist()
    label_word_fd = ConditionalFreqDist()
    for label, words in labeled_words:
        for word in words:
            word_fd[word] += 1
            label_word_fd[label][word] += 1
    n_xx = label_word_fd.N() # the total frequency for all words in all labels
    high_info_words = set()
    for label in label_word_fd.conditions():
        n_xi = label_word_fd[label].N() # the total frequency of all words that occurred for the label
        word_scores = collections.defaultdict(int)
        for word, n_ii in label_word_fd[label].items():
            n_ix = word_fd[word] # the total frequency of the word across all labels
            score = score_fn(n_ii, (n_ix, n_xi), n_xx)
            word_scores[word] = score
        bestwords = [word for word, score in word_scores.items() if score >= min_score]
        high_info_words |= set(bestwords)
    return high_info_words

#--------------------------------------------------------------

from nltk.corpus import movie_reviews

labels = movie_reviews.categories()

labeled_words = [(l,movie_reviews.words(categories=l)) for l in labels]

high_words = high_information_words(labeled_words)

lfeat = label_feats_from_corpus(movie_reviews,lambda words: bag_of_words(set(words) & set(high_words)))

train_feats_with,test_feats_with = split_label_feats(lfeat)

## SAMPLE CODE:

# nb_classifier = NaiveBayesClassifier.train(train_feats_with) : with high informative words
# accuracy(nb_classifier, test_feats_with)
# nb_precisions_with['pos'], nb_recalls_with['pos']
# nb_precisions_with['neg'], nb_recalls_with['neg']

<br>(precision,recall)

<b><font color="green">NaiveBayesClassifier</font></b>
<br>
<b>without</b>: 0.728
<br> pos : (0.651595744680851, 0.98)
<br> neg : (0.9596774193548387, 0.476)
<br><br>
<b>with</b>: 0.91
<br> pos : (0.8988326848249028, 0.924)
<br> neg : (0.9218106995884774, 0.896)
<br><br>
<b><font color="green">MaxentClassifier</font></b>
<br>
<b>without</b>: 0.722
<br> pos : (0.6456692913385826, 0.984)
<br> neg : (0.9663865546218487, 0.46)
<br><br>
<b>with</b>: 0.912
<br> pos : (0.8992248062015504, 0.928)
<br> neg : (0.9256198347107438, 0.896)
<br><br>
<b><font color="green">DecisionTreeClassifier</font></b>
<br>
<b>without</b>: 0.68
<br>
<b>with</b>: 0.706
<br><br>
<b><font color="green">SklearnClassifier</font></b>
<br>
<b>without</b>: 0.864
<br>
<b>with</b>: 0.86




In [20]:
from nltk.classify import NaiveBayesClassifier
nb_classifier = NaiveBayesClassifier.train(train_feats_with)
# accuracy(nb_classifier, test_feats_with)

from nltk.classify import MaxentClassifier
me_classifier = MaxentClassifier.train(train_feats_with, algorithm='gis', trace=0, max_iter=10, min_lldelta=0.5)

from nltk.classify import DecisionTreeClassifier
dt_classifier = DecisionTreeClassifier.train(train_feats_with, binary=True, depth_cutoff=5, 
                                             support_cutoff=30, entropy_cutoff=0.8)

from nltk.classify.scikitlearn import SklearnClassifier
from sklearn.svm import LinearSVC

sk_classifier = SklearnClassifier(LinearSVC()).train(train_feats_with)

In [31]:
#voting #vote #combining classifier with voting


import itertools
from nltk.classify import ClassifierI
from nltk.probability import FreqDist
class MaxVoteClassifier(ClassifierI):
    def __init__(self, *classifiers):
        self._classifiers = classifiers
        self._labels =sorted(set(itertools.chain(*[c.labels() for c in classifiers])))
    def labels():
        return self._labels
    def classify(self,feats):
        counts = FreqDist()
        for classifier in self._classifiers:
            counts[classifier.classify(feats)] +=1
        return counts.max()

        
# itertools.chain:
# [c.labels() for c in classifiers] == [['neg', 'pos'], ['pos', 'neg'],...]
# list(itertools.chain([c.labels() for c in classifiers])) == [['neg', 'pos'], ['pos', 'neg'],...]
# list(itertools.chain(*[c.labels() for c in classifiers])) == ['neg', 'pos','neg', 'pos',...]
# list(itertools.chain(*[c.labels() for c in classifiers])) == {'neg','pos'}


# itertools.chain('ab', 'cd', [1,2]) == a b c d 1 2 
# itertools.chain(['ab', 'cd', [1,2]]) == ab cd [1, 2]
# itertools.chain(*['ab', 'cd', [1,2]]) == a b c d 1 2


mv_classifier = MaxVoteClassifier(nb_classifier, dt_classifier, me_classifier, sk_classifier)
mv_classifier._labels


['neg', 'pos']

In [ ]:
#multilabel #multi label #multilabel classifier

from nltk.corpus import reuters
from nltk.metrics import BigramAssocMeasures
from collections import defaultdict
from sklearn.linear_model import LogisticRegression

# we need this format : (label, words)
def reuters_high_info_words(score_fn=BigramAssocMeasures.chi_sq):
    labels = reuters.categories()
    labeled_words = list()
    for label in labels:
        labeled_words.append((label,reuters.words(categories=label)))
    return high_information_words(labeled_words,score_fn)

#create test_feats and train_feats  format: (feats,labels)
def reuters_train_test_feats(feature_detector=bag_of_words):
    train_feats = list()
    test_feats = list()
    for fileid in reuters.fileids():
        feats = feature_detector(reuters.words(fileids=[fileid]))
        labels = reuters.categories(fileids=[fileid])
        if fileid.startswith('training'):
            train_feats.append((feats,labels))
        else:
            test_feats.append((feats,labels))
    return train_feats,test_feats
            
#get the words
hiws = reuters_high_info_words()
multi_train_feats,multi_test_feats = reuters_train_test_feats(lambda words:bag_of_words(set(hiws) & set(words)))


#train binary classification
def train_binary_classifiers(trainf, labelled_feats, labelset):
    pos = defaultdict(list)
    neg = defaultdict(list)
    classifiers = dict()
    
    for feats,labels in labelled_feats:
        for label in labels:
            pos[label].append((feats,label))
        for label in set(labelset)-set(labels):
            neg[label].append((feats,'!'+label))
    for label in labelset:
        classifiers[label] = trainf(pos[label]+neg[label])
    return classifiers



trainf = lambda train_feats:SklearnClassifier(LogisticRegression()).train(train_feats)
classifiers=train_binary_classifiers(trainf,multi_train_feats,reuters.categories())
len(classifiers)

In [ ]:
#multilabeled classifier

from nltk.classify import MultiClassifierI

class MultiBinaryClassifier(MultiClassifierI):
    def __init__(self,*classifiers):
        self._classifiers = dict(classifiers)
        self._labels = self._classifiers.keys()
    def labels():
        return self._labels
    def classify(self,feats):
        labels = set()
        for label,classifier in self._classifiers.items():
            if classifier.classify(feats)==label:
                labels.add(label)
        return labels


multi_classifier = MultiBinaryClassifier(*classifiers.items())

from nltk.metrics import masi_distance, precision, recall
def multi_metrics(multi_classifier, test_feats):
    mds = []
    refsets = collections.defaultdict(set)
    testsets = collections.defaultdict(set)
    for i, (feat, labels) in enumerate(test_feats):
        for label in labels:
            refsets[label].add(i)
        guessed = multi_classifier.classify(feat)
        for label in guessed:
            testsets[label].add(i)
        mds.append(masi_distance(set(labels), guessed))
    avg_md = sum(mds) / len(mds)
    precisions = {}
    recalls = {}
    for label in multi_classifier.labels():
        precisions[label] = precision(refsets[label], testsets[label])
        recalls[label] = recall(refsets[label], testsets[label])
    return precisions, recalls, avg_md





multi_precisions, multi_recalls, avg_md = multi_metrics(multi_classifier, multi_test_feats)

multi_precisions['soybean'], multi_recalls['soybean'], len(reuters.fileids(categories=['soybean']))


# Article Spinner


In [26]:
import nltk
import numpy as np
from bs4 import BeautifulSoup
from collections import defaultdict

positive_review = BeautifulSoup(open('/home/aminbag/nltk_data/sorted_data_acl/electronics/positive.review'),'lxml')
positive_review = positive_review.findAll('review_text')
positive_review = [review.get_text() for review in positive_review]


# create trigrams : { (W1,W3):[W2,W2-1,...],... }
trigrams = defaultdict(list)
for review in positive_review:
    review = review.lower()
    words = nltk.tokenize.word_tokenize(review)
    for i in range(len(words)-2):
        trigrams[(words[i],words[i+2])].append(words[i+1])


# محاسبه احتمال مربوط به هر یک از مقادیر
trigram_probabilities = {} # the values are dictionary from word->probability

for contex,center in trigrams.items():
    if len(set(center))>1:
        d = defaultdict(float)
        for word in center:
            d[word]+=1
        n = len(center)
        for word in d.keys():
            d[word] /= n
            
        trigram_probabilities[contex] = d


# انتخاب یک رندوم بر اساس وزن های کلمات
import random
def random_sample(d:dict):
    r = random.random()
    cum = 0
    
    for word,prob in d.items():
        cum +=prob
        if r<cum:
            return word

        


def test_spinner():
    review = random.choice(positive_review)
    s = review.lower()
    print('original: %s' %s)
    
    tokens = nltk.tokenize.word_tokenize(s)
    for i in range(len(tokens)-2):
        if random.random()<0.2: # 20% chance of replacement
            k = (tokens[i], tokens[i+2])
            if k in trigram_probabilities:
                w = random_sample(trigram_probabilities[k])
                tokens[i+1] = w
    print(" ".join(tokens)
          .replace(" .", ".").replace(" '", "'").replace(" ,", ",").replace("$ ", "$").replace(" !", "!"))
    


test_spinner()

original: 
this is truly a no-brainer, excellent headphones at a great price, wonderful all around phones, would make a great gift also

this is truly a no-brainer, excellent headphones at a great price, wonderful all the phones, would make a great gift also


In [18]:
import re
string = 'purple b@google.com amin@gmail.com monkey dishwasher'
match = re.findall(r'(\w+)@([\w.-]+)', string)
if match:
    print(match[0])

('b', 'google.com')
